### PR TITLE
Fix ftp traversal error conditions

### LINKED ISSUE #7582: Stack trace: colorado_ftp_traversal
## Steps to reproduce
setg RHOSTS file:test2/21.txt
setg THREADS 255
setg RPORT 21
use auxiliary/scanner/ftp/colorado_ftp_traversal
run

## Expected behavior
Show scan results.

## Current behavior
```
[*] 10.0.1.6:21           - 530 Please login with USER and PASS.
[-] Auxiliary failed: NoMethodError undefined method `read' for nil:NilClass
Did you mean?  rand
[-] Call stack:
[-]   /usr/share/metasploit-framework/modules/auxiliary/scanner/ftp/colorado_ftp_traversal.rb:75:in `run_host'
[-]   /usr/share/metasploit-framework/lib/msf/core/auxiliary/scanner.rb:135:in `block (2 levels) in run'
[-]   /usr/share/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
```
### Metasploit version
v4.12.42-dev 

### I installed Metasploit with:
Kali rolling

### OS
Kali rolling

### PR DESCRIPTION
**Pull request by h00die**
Fixes #7582 by checking if `sock` exists before sending data over it, or reading from it.  This is a combined solution by @h00die and @bwatters-r7 and was confirmed as a fix in the original ticket by @jinq102030.

## Verification

You'll want to run these against a NON-vulnerable version (like iis ftp, or vsftp)

- [ ] Start `msfconsole`
- [ ] `use modules/auxiliary/scanner/ftp/bison_ftp_traversal`
- [ ] set rhosts
- [ ] `scan`
- [ ] repeat for colorado, konica, pcman

### CODE REVIEW

File: modules/auxiliary/scanner/ftp/bison_ftp_traversal.rb [hunk 1]
```diff
@@ -61,32 +61,46 @@ def run_host(target_host)
...
+        file = ::File.basename(file_path)
+
+        # make RETR request and store server response message...
+        retr_cmd = ( "..//" * datastore['DEPTH'] ) + "#{file_path}"
+        res = send_cmd( ["RETR", retr_cmd])
+
+        # read the file data from the socket that we opened
+        # dont assume theres still a sock to read from. Per #7582
+        if sock.nil?
+          return
```
Comment (sempervictus) [inline modules/auxiliary/scanner/ftp/bison_ftp_traversal.rb]: Should we be tossing a log entry or user-facing message here?

Comment (bwatters-r7) [inline modules/auxiliary/scanner/ftp/bison_ftp_traversal.rb]: I am always a fan of logging.

**Comment by h00die** [comment]
did a copy/paste hack job on the throwing lines from @bwatters-r7 .

**Comment by bwatters-r7** [comment]
Paging @jinq102030; you are the one with the platform to test this.

**Comment by jinq102030** [comment]
Sorry, got distracted with another issue.  Will work on it this afternoon.  Thanks.

**Comment by jinq102030** [comment]
Verified the fix for all the 4 ftp modules.  Here is the run for bison_ftp
```
msf auxiliary(bison_ftp_traversal) > run

[*] 127.0.0.1:21          - /home/jqian/rapid7/metasploit-framework/modules/auxiliary/scanner/ftp/bison_ftp_traversal.rb::run_host:data_connect failed; posssible invalid response
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

**Comment by tdoan-r7** [comment]
# Release Notes 

This fix addresses an issue that caused errors when using FTP traversal modules. This fix checks if sock exists before it sends data over it or reads from it.

**Pull request merged**